tests/test_primitives/test_container/test_base_container/test_replace.py

```diff
@@ -0,0 +1,128 @@
+        """Dummy replace function for type checking."""
+        return container_instance
+
+    if not hasattr(copy, 'replace'):
+        copy.replace = _replace  # type: ignore
+
+
+class _CustomClass:
+    __slots__ = ('inner_value',)
+
+    def __init__(self, inner_value: str) -> None:
+        self.inner_value = inner_value
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, _CustomClass):
+            return self.inner_value == other.inner_value
+        return NotImplemented
+
+    def __ne__(self, other: object) -> bool:
+        if isinstance(other, _CustomClass):
+            return self.inner_value != other.inner_value
+        return NotImplemented
+
+    def __hash__(self) -> int:
+        return hash(self.inner_value)
+
+
+@given(
+    st.one_of(
+        st.integers(),
+        st.floats(allow_nan=False),
+        st.text(),
+        st.booleans(),
+        st.lists(st.text()),
+        st.dictionaries(st.text(), st.integers()),
+        st.builds(_CustomClass, st.text()),
+    ),
+)
+@example(None)
+def test_replace(container_value: Any) -> None:
+    """Test __replace__ magic method."""
+    container = BaseContainer(container_value)
+
+    new_value = 'new_value'
+    new_container = container.__replace__(_inner_value=new_value)
+
+    assert new_container is not container
+    assert new_container._inner_value == new_value  # noqa: SLF001
+    assert isinstance(new_container, BaseContainer)
+    assert type(new_container) is type(container)  # noqa: WPS516
+
+
+def test_replace_no_changes() -> None:
+    """Test __replace__ with no changes."""
+    container = BaseContainer('test')
+    result = container.__replace__()  # noqa: PLC2801
+    assert result is container
+
+
+def test_replace_invalid_attributes() -> None:
+    """Test __replace__ with invalid attributes."""
+    container = BaseContainer('test')
+
+    with pytest.raises(ValueError, match='Only _inner_value can be replaced'):
+        container.__replace__(invalid_attr='value')
+
+    with pytest.raises(ValueError, match='Only _inner_value can be replaced'):
+        container.__replace__(_inner_value='new', another_attr='value')
+
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 13),
+    reason='copy.replace requires Python 3.13+',
+)
+@given(
+    st.one_of(
+        st.integers(),
+        st.floats(allow_nan=False),
+        st.text(),
+        st.booleans(),
+        st.lists(st.text()),
+        st.dictionaries(st.text(), st.integers()),
+        st.builds(_CustomClass, st.text()),
+    ),
+)
+@example(None)
+def test_copy_replace(container_value: Any) -> None:
+    """Test copy.replace with BaseContainer."""
+    container = BaseContainer(container_value)
+
+    assert copy.replace(container) is container  # type: ignore[attr-defined]
+
+    new_value = 'new_value'
+    new_container = copy.replace(container, _inner_value=new_value)  # type: ignore[attr-defined]
```

Comment:
The problem with `copy.replace` is that it breaks typing :(

It does not know that `copy.replace(Success(1), 'a')` will be `Result[str, Any]` and not `Result[int, Any]`

We need a typed wrapper around it for our containers. With HKT support.